### PR TITLE
Modularize Earlystopping callback for easier subclassing

### PR DESCRIPTION
**Pull request by joelthchao**
### Summary

This PR allows user to easily subclass EarlyStopping to use their own metric calculation function.

### Sample usage
(modified examples/mnist_mlp.py)
```python
from __future__ import print_function

import keras
from keras.callbacks import EarlyStopping
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.optimizers import RMSprop

batch_size = 128
num_classes = 10
epochs = 20

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

model = Sequential()
model.add(Dense(512, activation='relu', input_shape=(784,)))
model.add(Dropout(0.2))
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(num_classes, activation='softmax'))

model.summary()

model.compile(loss='categorical_crossentropy',
              optimizer=RMSprop(),
              metrics=['accuracy'])

class MyEarlyStopping(EarlyStopping):
    def get_monitor_value(self, logs):
        return self.model.evaluate(x_test, y_test, verbose=False)[0]

callback = MyEarlyStopping(mode='min', patience=3)
history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    verbose=1,
                    validation_data=(x_test, y_test),
                    callbacks=[callback])

score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
```

### Related Issues

#10587 (feature request)

### PR Overview

- [ ] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [ ] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR. I don't think we should make this change to the API, because:

- It's not clear that it would be sufficiently useful to a sufficient number of users.
- A user could just subclass `EarlyStopping` to achieve the same functionality without too much trouble.

Maybe we should consider refactoring `EarlyStopping` to improve the UX of adding such functionality (via subclassing), without changing the API? For instance we could modularize into a private function the logic that gets the reference monitor value...

**Comment by fchollet** [comment]
For instance we could simply have a `get_monitor_value()` method that could be overridden.
